File: ch05/VB/Explorer/Explorer/ExplorerForm.vb
Imports System.IO

Public Class ExplorerForm
    Private Sub ExplorerForm_Load(sender As Object, e As EventArgs) Handles MyBase.Load
        ViewToolStripComboBox.SelectedIndex = 0

        ListView1.Columns.Add("Name", 250)
        ListView1.Columns.Add("Date Modified", 150)
        ListView1.Columns.Add("Size", 75, HorizontalAlignment.Right)

        Dim docs As New TreeNode("Raz Family")
        docs.Tag = Environment.GetFolderPath(Environment.SpecialFolder.UserProfile)
        TreeView1.Nodes.Add(docs)

        GetFolders(docs)
        docs.Expand()
    End Sub

    Private Sub GetFolders(node As TreeNode)
        Dim dir As New DirectoryInfo(node.Tag)

        Try
            For Each childDir In dir.GetDirectories()
                If childDir.Attributes.HasFlag(FileAttributes.Hidden) Then
                    Continue For
                End If

                Dim childNode As New TreeNode(childDir.Name)
                childNode.Tag = childDir.FullName
                node.Nodes.Add(childNode)

                GetFolders(childNode)
            Next
        Catch ex As Exception
            MsgBox(ex.Message)
        End Try

    End Sub

    Private Sub TreeView1_AfterSelect(sender As Object, e As TreeViewEventArgs) Handles TreeView1.AfterSelect
        Dim dir As New DirectoryInfo(e.Node.Tag)

        ListView1.Items.Clear()
        SmallImageList.Images.RemoveByKey(".exe")
        LargeImageList.Images.RemoveByKey(".exe")
        For Each thisFile In dir.GetFiles()
            Dim item As New ListViewItem(thisFile.Name)
            Dim lastWrite = thisFile.LastWriteTime
            item.SubItems.Add(lastWrite.ToShortDateString & " " & lastWrite.ToShortTimeString)
            item.SubItems.Add(Math.Ceiling(thisFile.Length / 1024) & " KB")

            If Not LargeImageList.Images.ContainsKey(thisFile.Extension) Then
                Dim thisIcon = Icon.ExtractAssociatedIcon(thisFile.FullName)
                SmallImageList.Images.Add(thisFile.Extension, thisIcon)
                LargeImageList.Images.Add(thisFile.Extension, thisIcon)
            End If

            item.ImageKey = thisFile.Extension
            ListView1.Items.Add(item)
        Next
    End Sub

    Private Sub ViewToolStripComboBox_SelectedIndexChanged(sender As Object, e As EventArgs) Handles ViewToolStripComboBox.SelectedIndexChanged
        Select Case ViewToolStripComboBox.Text
            Case "Large Icons"
                ListView1.View = View.LargeIcon
            Case "Small Icons"
                ListView1.View = View.SmallIcon
            Case "Details"
                ListView1.View = View.Details
        End Select
    End Sub
End Class

File: ch05/CS/Explorer/Explorer/ExplorerForm.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;

namespace Explorer
{
    public partial class ExplorerForm : Form
    {
        public ExplorerForm()
        {
            InitializeComponent();
        }

        private void ExplorerForm_Load(object sender, EventArgs e)
        {
            viewToolStripComboBox.SelectedIndex = 0;

            listView1.Columns.Add("Name", 250);
            listView1.Columns.Add("Last Modified", 150);
            listView1.Columns.Add("Size", 75, HorizontalAlignment.Right);

            var docs = new TreeNode(@"C:\users\Raz Family");
            docs.Tag = Environment.GetFolderPath(Environment.SpecialFolder.UserProfile);
            treeView1.Nodes.Add(docs);

            GetFolders(docs);
            docs.Expand();
        }

        private void GetFolders(TreeNode node)
        {
            var dir = new DirectoryInfo(node.Tag.ToString());
            try
            {
                foreach (var childDir in dir.GetDirectories())
                {
                    if (childDir.Attributes.HasFlag(FileAttributes.Hidden))
                    {
                        continue;
                    }
                    var childNode = new TreeNode(childDir.Name);
                    childNode.Tag = childDir.FullName;
                    node.Nodes.Add(childNode);

                    GetFolders(childNode);
                }
            }
            catch (Exception e)
            {
                MessageBox.Show(e.Message);
            }

        }

        private void treeView1_AfterSelect(object sender, TreeViewEventArgs e)
        {
            var dir = new DirectoryInfo(e.Node.Tag.ToString());

            listView1.Items.Clear();
            smallImageList.Images.RemoveByKey(".exe");
            largeImageList.Images.RemoveByKey(".exe");
            foreach (var thisFile in dir.GetFiles())
            {
                var item = new ListViewItem(thisFile.Name);
                var lastWrite = thisFile.LastWriteTime.ToShortDateString() + " " + thisFile.LastWriteTime.ToShortTimeString();
                item.SubItems.Add(lastWrite);
                item.SubItems.Add(Math.Ceiling(thisFile.Length / 1024.0) + " KB");

                if (! largeImageList.Images.ContainsKey(thisFile.Extension))
                {
                    var thisIcon = Icon.ExtractAssociatedIcon(thisFile.FullName);
                    smallImageList.Images.Add(thisFile.Extension, thisIcon);
                    largeImageList.Images.Add(thisFile.Extension, thisIcon);
                }
                item.ImageKey = thisFile.Extension;
                listView1.Items.Add(item);
            }
        }

        private void viewToolStripComboBox_SelectedIndexChanged(object sender, EventArgs e)
        {
            switch (viewToolStripComboBox.Text)
            {
                case "Large Icons":
                    listView1.View = View.LargeIcon;
                    break;
                case "Small Icons":
                    listView1.View = View.SmallIcon;
                    break;
                case "Details":
                    listView1.View = View.Details;
                    break;
                default:
                    break;
            }
        }
    }
}
